docs/type-constraints-on-generic-parameters.md

```diff
@@ -0,0 +1,216 @@
+
+This is the main reasoning behind this modification. Many other languages have a feature to limit what types does a generic allow as a baseline. Languages like C# have this implemented on the form of the `where T : BaseClass` syntax, which delimits the type that `T` can be or inherit from. It supports cases in which you may want to use a strucutre, but not know its implementation details, like an interface.
+
+However, we can use this pattern to improve the reliability of code, a function such as the following that takes a string and returns a type, without having to type cast into `any` before type casting into `T`.
+
+
+```luau
+type GameService = {
+    -- ...
+}
+
+type GameManager = {
+    Disconnect: (self: GameManager) -> (),
+    -- ...
+} & GameService
+
+type ServiceContainer = {
+    ServiceList: { [string]: GameService },
+    Require: <T: GameService>(self: ServiceContainer, serviceName: string) -> T
+}
+
+local ServiceContainer = {
+    Require = function<T: GameService>(self: ServiceContainer, serviceName: string): T 
+        return self.ServiceList[serviceName] :: T
+    end
+}
+
+-- ...
+
+local GameManager = ServiceContainer:Require<GameManager>("@game/GameManager")
+GameManager:Disconnect()
+```
+
+## Design
+
+The design of this feature will reuse the already existing generics syntax, with little modification.
+
+```luau
+type Entity = {
+    -- ...
+}
+
+type Gun = {
+    IsEmpty: (self: Gun) -> boolean,
+    FireGun: (self: Gun) -> (),
+    -- ...
+}
+
+type WeaponResult = {
+    HittedEntity: Entity,
+    DamageDealt: number,
+}
+
+type ShotgunWeaponResult =  {
+    PelletCount: number,
+    AmountOfPelletsHit: number,
+} & WeaponResult
+
+type Shotgun = {
+    -- ...
+} & Gun
+
+local function fireGun<T: Gun, U: WeaponResult>(self: T): U?
+    if not self:IsEmpty() then        -- No error
+        return self:FireGun()         -- No error
+    end
+    
+    return nil    
+end 
+
+local gun: Gun
+local shotgun: Shotgun
+
+local fireResult = fireGun(gun) :: WeaponResult                     -- No error
+local shotgunFireResult = fireGun(shotgun) :: ShotgunWeaponResult   -- No error
+
+fireGun("Hello, world!")     -- Type Error: type 'string' and 'Gun' are not related to one another.
+```
+
+The syntax for defining a generic function would remain the same, except, instead of simply taking a `T`, we would take a `T : TypeConstraint`, the syntax remains similar to other features of the language, and it would work by checking if the type that `T` is inferred to be has any kind of relationship with `TypeConstraint`. However, we may also want to be explicit when we are doing generics on this way, therefore a syntax to properly and explicitly specify which types the function takes in its current call would be preferred, this way, we no longer have to use turbofish to type cast into the correct return type `U` constrained by `WeaponResult`.
```

Comment:
I don't know why you're mentioning turbofish, that doesn't exist in Luau, and can't because it's ambiguous with `:: <T>() -> ()`.

No turbofish is also not an option for the exact same reasons it's necessary in Rust.

```luau
local a, b = the<guardian, stands>(resolute)
```

...is perfectly valid code today.

Reply:
I remember discussing with Drak that `f.<T, U>()` isn't the worst syntax in the world, but also it feels out of scope for this RFC anyway?

Reply:
You're right, it feels out of scope for this RFC to add on about, and it was already somewhat addressed on the original generic functions rfc

Reply:
although I did not know that that snippet was valid ever, however, all I keep getting from attempting it is syntax errors